crypto/rsa_extra/rsa_test.cc

```diff
@@ -1214,70 +1193,33 @@ TEST(RSATest, KeygenFailOnce) {
 }
 
 #else
+
 // AWSLCAndroidTestRunner does not take tests that do |ASSERT_DEATH| very well.
 // GTEST issue: https://github.com/google/googletest/issues/1496.
 #if !defined(OPENSSL_ANDROID)
 
-// In the case of a FIPS build, expect abort() when |RSA_generate_key_ex| fails.
-TEST(RSADeathTest, KeygenFailAndDie) {
-  bssl::UniquePtr<RSA> rsa(RSA_new());
-  ASSERT_TRUE(rsa);
-
-  // Cause RSA key generation after a prime has been generated, to test that
-  // |rsa| is left alone.
-  BN_GENCB cb;
-  BN_GENCB_set(&cb,
-               [](int event, int, BN_GENCB *) -> int { return event != 3; },
-               nullptr);
-
-  bssl::UniquePtr<BIGNUM> e(BN_new());
-  ASSERT_TRUE(e);
-  ASSERT_TRUE(BN_set_word(e.get(), RSA_F4));
-
-  // Key generation should fail.
-  ASSERT_DEATH_IF_SUPPORTED(RSA_generate_key_ex(rsa.get(), 2048, e.get(), &cb),"");
-
-  // Failed key generations do not leave garbage in |rsa|.
-  EXPECT_FALSE(rsa->n);
-  EXPECT_FALSE(rsa->e);
-  EXPECT_FALSE(rsa->d);
-  EXPECT_FALSE(rsa->p);
-  EXPECT_FALSE(rsa->q);
-  EXPECT_FALSE(rsa->dmp1);
-  EXPECT_FALSE(rsa->dmq1);
-  EXPECT_FALSE(rsa->iqmp);
-  EXPECT_FALSE(rsa->mont_n);
-  EXPECT_FALSE(rsa->mont_p);
-  EXPECT_FALSE(rsa->mont_q);
-  EXPECT_FALSE(rsa->d_fixed);
-  EXPECT_FALSE(rsa->dmp1_fixed);
-  EXPECT_FALSE(rsa->dmq1_fixed);
-  EXPECT_FALSE(rsa->iqmp_mont);
-  EXPECT_FALSE(rsa->private_key_frozen);
-
-  // Failed key generations leave the previous contents alone.
-  EXPECT_TRUE(RSA_generate_key_ex(rsa.get(), 2048, e.get(), nullptr));
-  uint8_t *der;
-  size_t der_len;
-  ASSERT_TRUE(RSA_private_key_to_bytes(&der, &der_len, rsa.get()));
-  bssl::UniquePtr<uint8_t> delete_der(der);
+static bool set_env(const std::string& name, const std::string& value) {
+#if defined(OPENSSL_WINDOWS)
+  return SetEnvironmentVariableA(name.c_str(), value.c_str()) != 0;
+#else
+  return setenv(name.c_str(), value.c_str(), 1) == 0;
+#endif
+}
 
-  ASSERT_DEATH_IF_SUPPORTED(RSA_generate_key_ex(rsa.get(), 2048, e.get(), &cb),"");
+// In the case of a FIPS build, expect abort() when |RSA_generate_key_fips|
+// fails.
+TEST(RSADeathTest, KeygenFailAndDie) {
```

Comment:
What about this test would fail, why is this a valuable test? I think this is causing a bunch of CI failures too. 

Previously was the BN callback the reason the keygen would fail? What about the test setup indicates the PCT would fail if you are using supported key lengths and doing normal keygen?

Reply:
CI failures are on me, I think the environment variables aren’t being set properly in the CI environment. I'll change the env variable to be set in the script and skip the test if it's not defined.

> Why is this a valuable test?

FIPS has requirements for Pairwise Consistency Tests to be implemented in `RSA_check_fips`. If anything goes wrong in these specific consistency tests, the library should be aborted. This test was checking if the library was aborted correctly. Originally we were aborting for **any general** RSA key generation failure, this changes the behavior to abort only when the PCT tests fail.

> Previously was the BN callback the reason the keygen would fail? What about the test setup indicates the PCT would fail if you are using supported key lengths and doing normal keygen?

Yeah the callback was forcing a general failure in keygen. This was unrelated to PCT however.
In normal circumstances, the PCT can't fail unless something goes horribly wrong with the module. Testing that the module properly aborts is required for FIPS demonstrations, which is why we have existing logic to force RSA PCT failures.
https://github.com/aws/aws-lc/blob/59828538a790094113eacd5dd23d01be2885b36a/crypto/fipsmodule/rsa/rsa.c#L1453-L1455

The new test set up leverages the existing logic, so we properly test for PCT aborting failures.

Reply:
Ah I see that makes sense. That last link is helpful - I was wondering how the failure was being forced. Thanks!